server/integration/testsuite/src/test/java/org/infinispan/server/test/partitionhandling/PartitionHandlingIT.java

```diff
@@ -0,0 +1,257 @@
+      assertEquals("value2", allowReadWritesCache3.get(server1OwnedKey));
+
+      healCluster();
+   }
+
+   /**
+    * This check exists because the rebalance happens after the view change, and if any test runs in this window the getKeyOwnedByNode
+    * method may end in an infinite loop. The rebalance status could be checked in PartitionHandlingController using creeper, but for an
+    * unknown reason that doesn't work, so this is sort of a workaround.
+    */
+   private boolean assertNoRebalance(String cacheName, RemoteInfinispanServer... servers) {
+      return Arrays.stream(servers).allMatch(
+            server -> ITestUtils.createCacheManager(server).getCache(cacheName).getCacheTopologyInfo().getSegmentsPerServer().size() == servers.length);
+   }
+
+   private void partitionCluster() {
+      partitionHandlingController.partitionCluster(new HashSet<>(Arrays.asList("node0", "node1")),
+                                                   new HashSet<>(Arrays.asList("node2")));
+   }
+
+   private void healCluster() {
+      partitionHandlingController.heal();
+   }
+
+   private void assertThrowsHotRodClientException(Runnable r) {
+      try {
+         r.run();
+         fail();
+      } catch (HotRodClientException hce){
+         // OK
+      }
+   }
+
+   private String getKeyOwnedByNode(String serverAddress, RemoteCacheManager rcm, RemoteCache cache, int owners) {
+      SegmentConsistentHash hash = createHash(cache, NUM_SEGMENTS, owners);
+
+      int segmentOwnedByNode = -1;
+      for (Map.Entry<SocketAddress, Set<Integer>> socketAddressToSegments : hash.getSegmentsByServer().entrySet()) {
+         if (socketAddressToSegments.getKey().toString().equals(serverAddress)) {
+            segmentOwnedByNode = socketAddressToSegments.getValue().stream().findFirst().get();
+         }
+      }
+
+      int i = 0;
+      while (hash.getSegment(marshall(rcm, "key" + i)) != segmentOwnedByNode) {
+         i++;
+         if (i > 10000) { // don't loop forever
+            throw new IllegalStateException("Server " + serverAddress + " is not an owner of any key.");
+         }
+      }
+      return "key" + i;
+   }
+
+   private SegmentConsistentHash createHash(RemoteCache cache, int segments, int owners) {
+      SocketAddress[][] mySegmentOwners = new SocketAddress[segments][owners];
+      for (Map.Entry<SocketAddress, Set<Integer>> socketAddressSetEntry : cache.getCacheTopologyInfo().getSegmentsPerServer().entrySet()) {
+         SocketAddress server = socketAddressSetEntry.getKey();
+         for (Integer integer : socketAddressSetEntry.getValue()) {
+            int emptyIndex = 0;
+            while (mySegmentOwners[integer][emptyIndex] != null) {
+               emptyIndex++;
+            }
+            mySegmentOwners[integer][emptyIndex] = server;
+         }
+      }
+      SegmentConsistentHash result = new SegmentConsistentHash();
+      result.init(mySegmentOwners, segments);
+      return result;
```

Comment:
Kind of annoying that you need all of this just to call `getSegment(key)`... wouldn't something like this be better? 

`((TcpTransportFactory) ((InternalRemoteCacheManager) remoteCacheManager).getTransportFactory()).getConsistentHash(RemoteCacheManager.cacheNameBytes(cache.getName()))` instead